File: vkyc-sdk/vkyc-ios-sdk/Examples/Objective-C/ViewController.m
//
//  ViewController.m
//  VKYC Example (Objective-C)
//
//  Created on 21/03/2026.
//  Copyright © 2026 VKYC. All rights reserved.
//

#import "ViewController.h"

#if __has_include(<VKYC/VKYC-Swift.h>)
#import <VKYC/VKYC-Swift.h>
#define VKYC_SDK_AVAILABLE 1
#elif __has_include("VKYC-Swift.h")
#import "VKYC-Swift.h"
#define VKYC_SDK_AVAILABLE 1
#else
#define VKYC_SDK_AVAILABLE 0
#endif

#if VKYC_SDK_AVAILABLE

@interface ViewController () <VKYCDelegate>
@property (nonatomic, strong) UIButton *startButton;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setupUI];
}

- (void)setupUI {
    self.view.backgroundColor = [UIColor whiteColor];
    self.title = @"VKYC Example (Obj-C)";
    
    self.startButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [self.startButton setTitle:@"Start VKYC" forState:UIControlStateNormal];
    self.startButton.titleLabel.font = [UIFont systemFontOfSize:18 weight:UIFontWeightSemibold];
    [self.startButton addTarget:self action:@selector(startVKYCTapped) forControlEvents:UIControlEventTouchUpInside];
    self.startButton.translatesAutoresizingMaskIntoConstraints = NO;
    
    [self.view addSubview:self.startButton];
    
    [NSLayoutConstraint activateConstraints:@[
        [self.startButton.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
        [self.startButton.centerYAnchor constraintEqualToAnchor:self.view.centerYAnchor]
    ]];
}

#pragma mark - Example 1: Basic Integration

- (void)startVKYCTapped {
    [self startOVSE];
}

- (void)startOVSE {
    VKYCConfig *config = [[VKYCConfig alloc] initWithToken:@""
                                                    apiKey:@"YOUR_GRIDLINES_API_KEY"
                                               environment:VKYCConfigEnvironmentStaging
                                                      mode:VKYCConfigModeOvse];

    VKYCOVSEOptions *ovse = [[VKYCOVSEOptions alloc] initWithApiBaseUrl:@"https://api-dev.gridlines.io/uidai-api/ovse"
                                                                  apiKey:@"YOUR_GRIDLINES_API_KEY"
                                                           initialApiKey:@"YOUR_GRIDLINES_API_KEY"
                                                              channelType:@"APP"
                                                               templateId:@"1"
                                                      expiryTimeInSeconds:@3600
                                                                  consent:@"Y"
                                                             appPackageId:@"in.ongrid.lav"
                                                             appSignature:@"DZ54P8HK5D"
                                                        pollingIntervalMs:@5000
                                                          maxPollAttempts:@60];
    config.ovse = ovse;

    VKYCTextOptions *texts = [[VKYCTextOptions alloc] initWithWelcomeTitle:@"Aadhaar OVSE"
                                                            welcomeSubtitle:@"Secure verification powered by OnGrid"
                                                            startButtonLabel:@"Start OVSE"
                                                        startOVSEButtonLabel:nil
                                                           cancelButtonLabel:nil
                                                                  ovseTitle:@"Aadhaar OVSE Verification"
                                                               ovseSubtitle:nil
                                                             ovseInputLabel:@"API Key"
                                                       ovseInputPlaceholder:nil
                                                            ovseSubmitLabel:@"Start Flow"];
    config.texts = texts;

    [VKYCManager startFrom:self
                    config:config
                   success:^(NSDictionary<NSString *,id> * _Nonnull result) {
        [self handleSuccess:result];
    }
                   failure:^(VKYCError * _Nonnull error) {
        [self handleError:error];
    }
                    cancel:^{
        [self handleCancellation];
    }];
}

#pragma mark - Example 2: With Custom Theme

- (void)startVKYCWithCustomTheme {
    // Create configuration
    VKYCConfig *config = [[VKYCConfig alloc] initWithToken:@"your-token"
                                                    apiKey:@"your-api-key"
                                               environment:VKYCConfigEnvironmentProduction];
    
    // Customize theme
    VKYCTheme *theme = [[VKYCTheme alloc] initWithPrimaryColor:@"#667eea"
                                                secondaryColor:@"#764ba2"
                                                     textColor:@"#333333"
                                           backgroundColor:@"#FFFFFF"
                                                    fontFamily:@"SF Pro Display"
                                                  buttonRadius:@12];
    config.theme = theme;
    
    // Configure features
    VKYCFeatures *features = [[VKYCFeatures alloc] initWithVideoEnabled:YES
                                                           autoCapture:@YES
                                                         livenessCheck:@YES
                                              documentVerification:@YES
                                                          faceMatch:@YES
                                                       audioEnabled:@YES
                                                   screenRecording:@NO];
    config.features = features;
    
    // Add metadata
    config.metadata = @{
        @"userId": @"12345",
        @"source": @"ios_app",
        @"sessionType": @"kyc_verification"
    };
    
    // Start VKYC
    [VKYCManager startFrom:self
                    config:config
                   success:^(NSDictionary<NSString *,id> * _Nonnull result) {
        [self handleSuccess:result];
    }
                   failure:^(VKYCError * _Nonnull error) {
        [self handleError:error];
    }
                    cancel:^{
        [self handleCancellation];
    }];
}

#pragma mark - Example 3: Using Delegate Pattern

- (void)startVKYCWithDelegate {
    VKYCConfig *config = [[VKYCConfig alloc] initWithToken:@"your-token"
                                                    apiKey:@"your-api-key"
                                               environment:VKYCConfigEnvironmentStaging];
    
    VKYCManager.shared.delegate = self;
    [VKYCManager startFrom:self config:config];
}

#pragma mark - Example 4: Validation Before Starting

- (void)startVKYCWithValidation {
    VKYCConfig *config = [[VKYCConfig alloc] initWithToken:@"your-token"
                                                    apiKey:@"your-api-key"
                                               environment:VKYCConfigEnvironmentStaging];
    
    // Validate configuration
    VKYCConfigValidationResult validationResult = [config validate];
    
    if (validationResult == VKYCConfigValidationResultSuccess) {
        [VKYCManager startFrom:self
                        config:config
                       success:^(NSDictionary<NSString *,id> * _Nonnull result) {
            [self handleSuccess:result];
        }
                       failure:^(VKYCError * _Nonnull error) {
            [self handleError:error];
        }
                        cancel:^{
            [self handleCancellation];
        }];
    } else {
        [self showAlertWithTitle:@"Invalid Configuration" message:@"Please check your configuration"];
    }
}

#pragma mark - Helper Methods

- (void)handleSuccess:(NSDictionary *)result {
    NSLog(@"✅ VKYC Success: %@", result);
    
    NSString *sessionId = result[@"sessionId"] ?: @"N/A";
    NSString *status = result[@"status"] ?: @"N/A";
    
    NSString *message = [NSString stringWithFormat:@"Session ID: %@\nStatus: %@", sessionId, status];
    [self showAlertWithTitle:@"Verification Successful" message:message];
}

- (void)handleError:(VKYCError *)error {
    NSLog(@"❌ VKYC Error: %@", error);
    
    NSString *message;
    switch (error.code) {
        case VKYCErrorCodeCameraPermissionDenied:
            message = @"Camera permission is required. Please grant access in Settings.";
            break;
        case VKYCErrorCodeMicrophonePermissionDenied:
            message = @"Microphone permission is required. Please grant access in Settings.";
            break;
        case VKYCErrorCodeNetworkError:
            message = @"Network error. Please check your connection and try again.";
            break;
        case VKYCErrorCodeSessionExpired:
            message = @"Your session has expired. Please start again.";
            break;
        default:
            message = error.message;
            break;
    }
    
    [self showAlertWithTitle:@"Verification Failed" message:message];
}

- (void)handleCancellation {
    NSLog(@"⚠️ VKYC Cancelled");
    [self showAlertWithTitle:@"Cancelled" message:@"Verification was cancelled"];
}

- (void)showAlertWithTitle:(NSString *)title message:(NSString *)message {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:title
                                                                   message:message
                                                            preferredStyle:UIAlertControllerStyleAlert];
    [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
    [self presentViewController:alert animated:YES completion:nil];
}

#pragma mark - VKYCDelegate

- (void)vkycDidStart {
    NSLog(@"📱 VKYC Started");
}

- (void)vkycDidSucceedWith:(NSDictionary<NSString *,id> *)result {
    NSLog(@"✅ VKYC Success: %@", result);
    [self handleSuccess:result];
}

- (void)vkycDidFailWith:(VKYCError *)error {
    NSLog(@"❌ VKYC Error: %@", error);
    [self handleError:error];
}

- (void)vkycDidCancel {
    NSLog(@"⚠️ VKYC Cancelled");
    [self handleCancellation];
}

- (void)vkycDidReceiveEvent:(NSString *)eventName metadata:(NSDictionary<NSString *,id> *)metadata {
    NSLog(@"📊 VKYC Event: %@, metadata: %@", eventName, metadata);
    
    // Track events with your analytics service
    // [Analytics track:eventName properties:metadata];
}

@end

#else

@implementation ViewController

@end

#endif

File: vkyc-sdk/vkyc-ios-sdk/Examples/Swift/ViewController.swift
//
//  ViewController.swift
//  VKYC Example
//
//  Created on 21/03/2026.
//  Copyright © 2026 VKYC. All rights reserved.
//

import UIKit
import VKYC

class ViewController: UIViewController {
    
    // MARK: - Properties
    
    private let startButton: UIButton = {
        let button = UIButton(type: .system)
        button.setTitle("Start VKYC", for: .normal)
        button.titleLabel?.font = .systemFont(ofSize: 18, weight: .semibold)
        button.translatesAutoresizingMaskIntoConstraints = false
        return button
    }()
    
    // MARK: - Lifecycle
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupUI()
    }
    
    private func setupUI() {
        view.backgroundColor = .white
        title = "VKYC Example"
        
        view.addSubview(startButton)
        startButton.addTarget(self, action: #selector(startVKYCTapped), for: .touchUpInside)
        
        NSLayoutConstraint.activate([
            startButton.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            startButton.centerYAnchor.constraint(equalTo: view.centerYAnchor)
        ])
    }
    
    // MARK: - Example 1: Basic Integration with Completion Handler
    
    @objc private func startVKYCTapped() {
        // Create configuration
        let config = VKYCConfig(
            token: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9...",
            apiKey: "your-api-key-here",
            environment: .staging
        )
        
        // Start VKYC
        VKYCManager.start(from: self, config: config) { result in
            switch result {
            case .success(let data):
                self.handleSuccess(data)
                
            case .failure(let error):
                self.handleError(error)
                
            case .cancelled:
                self.handleCancellation()
            }
        }
    }
    
    // MARK: - Example 2: Advanced Integration with Custom Theme
    
    private func startVKYCWithCustomTheme() {
        // Create configuration
        let config = VKYCConfig(
            token: "your-token",
            apiKey: "your-api-key",
            environment: .production
        )
        
        // Customize theme
        config.theme = VKYCTheme(
            primaryColor: "#667eea",
            secondaryColor: "#764ba2",
            textColor: "#333333",
            backgroundColor: "#FFFFFF",
            fontFamily: "SF Pro Display",
            buttonRadius: 12
        )
        
        // Configure features
        config.features = VKYCFeatures(
            videoEnabled: true,
            autoCapture: true,
            livenessCheck: true,
            documentVerification: true,
            faceMatch: true,
            audioEnabled: true,
            screenRecording: false
        )
        
        // Add custom metadata
        config.metadata = [
            "userId": "12345",
            "source": "ios_app",
            "sessionType": "kyc_verification"
        ]
        
        // Start VKYC
        VKYCManager.start(from: self, config: config) { result in
            self.handleResult(result)
        }
    }
    
    // MARK: - Example 3: Using Delegate Pattern
    
    private func startVKYCWithDelegate() {
        let config = VKYCConfig(
            token: "your-token",
            apiKey: "your-api-key",
            environment: .staging
        )
        
        VKYCManager.shared.delegate = self
        VKYCManager.start(from: self, config: config)
    }
    
    // MARK: - Example 4: Validation Before Starting
    
    private func startVKYCWithValidation() {
        let config = VKYCConfig(
            token: "your-token",
            apiKey: "your-api-key",
            environment: .staging
        )
        
        // Validate configuration
        switch config.validate() {
        case .success:
            VKYCManager.start(from: self, config: config) { result in
                self.handleResult(result)
            }
            
        case .failure(let errors):
            showAlert(
                title: "Invalid Configuration",
                message: errors.joined(separator: "\n")
            )
        }
    }
    
    // MARK: - Example 5: Checking Permissions First
    
    private func startVKYCWithPermissionCheck() {
        checkPermissions { [weak self] granted in
            guard let self = self else { return }
            
            if granted {
                let config = VKYCConfig(
                    token: "your-token",
                    apiKey: "your-api-key",
                    environment: .staging
                )
                
                VKYCManager.start(from: self, config: config) { result in
                    self.handleResult(result)
                }
            } else {
                self.showPermissionDeniedAlert()
            }
        }
    }
    
    // MARK: - Helper Methods
    
    private func handleResult(_ result: VKYCResult) {
        switch result {
        case .success(let data):
            handleSuccess(data)
        case .failure(let error):
            handleError(error)
        case .cancelled:
            handleCancellation()
        }
    }
    
    private func handleSuccess(_ data: [String: Any]) {
        print("✅ VKYC Success: \(data)")
        
        let sessionId = data["sessionId"] as? String ?? "N/A"
        let status = data["status"] as? String ?? "N/A"
        
        showAlert(
            title: "Verification Successful",
            message: "Session ID: \(sessionId)\nStatus: \(status)"
        )
    }
    
    private func handleError(_ error: VKYCError) {
        print("❌ VKYC Error: \(error)")
        
        let message: String
        switch error.code {
        case .cameraPermissionDenied:
            message = "Camera permission is required. Please grant access in Settings."
        case .microphonePermissionDenied:
            message = "Microphone permission is required. Please grant access in Settings."
        case .networkError:
            message = "Network error. Please check your connection and try again."
        case .sessionExpired:
            message = "Your session has expired. Please start again."
        default:
            message = error.message
        }
        
        showAlert(title: "Verification Failed", message: message)
    }
    
    private func handleCancellation() {
        print("⚠️ VKYC Cancelled")
        showAlert(title: "Cancelled", message: "Verification was cancelled")
    }
    
    private func checkPermissions(completion: @escaping (Bool) -> Void) {
        // Check camera permission
        let cameraStatus = AVCaptureDevice.authorizationStatus(for: .video)
        
        switch cameraStatus {
        case .authorized:
            completion(true)
            
        case .notDetermined:
            AVCaptureDevice.requestAccess(for: .video) { granted in
                DispatchQueue.main.async {
                    completion(granted)
                }
            }
            
        case .denied, .restricted:
            completion(false)
            
        @unknown default:
            completion(false)
        }
    }
    
    private func showPermissionDeniedAlert() {
        let alert = UIAlertController(
            title: "Permission Required",
            message: "Camera and microphone access is required for video verification. Please grant access in Settings.",
            preferredStyle: .alert
        )
        
        alert.addAction(UIAlertAction(title: "Settings", style: .default) { _ in
            if let settingsURL = URL(string: UIApplication.openSettingsURLString) {
                UIApplication.shared.open(settingsURL)
            }
        })
        
        alert.addAction(UIAlertAction(title: "Cancel", style: .cancel))
        
        present(alert, animated: true)
    }
    
    private func showAlert(title: String, message: String) {
        let alert = UIAlertController(
            title: title,
            message: message,
            preferredStyle: .alert
        )
        alert.addAction(UIAlertAction(title: "OK", style: .default))
        present(alert, animated: true)
    }
}

// MARK: - VKYCDelegate

extension ViewController: VKYCDelegate {
    
    func vkycDidStart() {
        print("📱 VKYC Started")
    }
    
    func vkycDidSucceed(with result: [String : Any]) {
        print("✅ VKYC Success: \(result)")
        handleSuccess(result)
    }
    
    func vkycDidFail(with error: VKYCError) {
        print("❌ VKYC Error: \(error)")
        handleError(error)
    }
    
    func vkycDidCancel() {
        print("⚠️ VKYC Cancelled")
        handleCancellation()
    }
    
    func vkycDidReceiveEvent(_ eventName: String, metadata: [String : Any]?) {
        print("📊 VKYC Event: \(eventName), metadata: \(metadata ?? [:])")
        
        // Track events with your analytics service
        // Analytics.track(eventName, properties: metadata)
    }
}

// MARK: - Import for permissions
import AVFoundation
